Composer/packages/lib/bot-deploy/src/botProjectDeploy.ts

```diff
@@ -0,0 +1,597 @@
+      publishingPassword: userPwd,
+    });
+    this.logger(updateRes);
+
+    const publishEndpoint = `https://${name}-${env}.scm.azurewebsites.net/zipdeploy`;
+
+    const publishCreds = Buffer.from(`${userName}:${userPwd}`).toString('base64');
+
+    const fileContent = await fs.readFile(zipPath);
+    const options = {
+      body: fileContent,
+      encoding: null,
+      headers: {
+        Authorization: `Basic ${publishCreds}`,
+        'Content-Type': 'application/zip',
+        'Content-Length': fileContent.length,
+      },
+    } as rp.RequestPromiseOptions;
+    const response = await rp.post(publishEndpoint, options);
+    this.logger(response);
+  }
+
+  public async create(
+    name: string,
+    location: string,
+    environment: string,
+    appPassword: string,
+    luisAuthoringKey?: string
+  ) {
+    const credsForGraph = new msRestNodeAuth.DeviceTokenCredentials(
+      this.creds.clientId,
+      this.tenantId,
+      this.creds.username,
+      'graph',
+      this.creds.environment,
+      this.creds.tokenCache
+    );
+    const graphClient = new GraphRbacManagementClient(credsForGraph, this.tenantId, {
+      baseUri: 'https://graph.windows.net',
+    });
+
+    if (!fs.existsSync(this.deploymentSettingsPath)) {
+      this.logger(`! Could not find an 'appsettings.deployment.json' file in the current directory.`);
+      return;
+    }
+
+    const settings = await fs.readJson(this.deploymentSettingsPath);
+    let appId = settings.MicrosoftAppId;
+
+    if (!appId) {
+      if (!appPassword) {
+        this.logger(`App password is required`);
+        return;
+      }
+      this.logger('> Creating App Registration ...');
+      const appCreated = await this.createApp(graphClient, name, appPassword);
+      this.logger(appCreated);
+      appId = appCreated.appId;
+    }
+
+    this.logger(`> Create App Id Success! ID: ${appId}`);
+
+    var shouldCreateAuthoringResource = true;
+    if (luisAuthoringKey) {
+      shouldCreateAuthoringResource = false;
+    }
+
+    const resourceGroupName = `${name}-${environment}`;
+
+    // timestamp will be used as deployment name
+    const timeStamp = new Date().getTime().toString();
+    const client = new ResourceManagementClient(this.creds, this.subId);
+
+    const rpres = await this.createResourceGroup(client, location, resourceGroupName);
+    this.logger(rpres);
+
+    const deploymentTemplateParam = this.getDeploymentTemplateParam(
+      appId,
+      appPassword,
+      location,
+      name,
+      shouldCreateAuthoringResource
+    );
+    this.logger(deploymentTemplateParam);
+
+    const validation = await this.validateDeployment(
+      client,
+      this.templatePath,
+      location,
+      resourceGroupName,
+      timeStamp,
+      deploymentTemplateParam
+    );
+    this.logger(validation);
+
+    if (validation.error) {
+      this.logger(`! Template is not valid with provided parameters. Review the log for more information.`);
```

Comment:
the interface for this component is not exposed at a fine enough grain. Let's update it. I see at least three statuses, perhaps 4:

`INFO`, `ERROR`, `WARNING`, `SUCCESS`

So I recommend an object that takes at least two fields, `status` & `message`

```javascript
this.logger({
  status: 'ERROR'
  message: ```
    Operation failed. Template is not valid with provided parameters. Review the log for more information. 
    To delete this resource group, run 'az group delete -g ${resourceGroupName} --no-wait'
    ```
})
```

Reply:
resolved